crates/ty_python_semantic/src/diagnostic.rs

```diff
@@ -144,3 +149,114 @@ where
 
     buffer
 }
+
+/// An abstraction for mutating a diagnostic.
+///
+/// Callers likely should use `LintDiagnosticGuard` via
+/// `InferContext::report_lint` instead. This guard is only intended for use
+/// with non-lint diagnostics or non-type checking diagnostics. It is fundamentally lower level and easier to
+/// get things wrong by using it.
+///
+/// Unlike `LintDiagnosticGuard`, this API does not guarantee that the
+/// constructed `Diagnostic` not only has a primary annotation, but its
+/// associated file is equivalent to the file being type checked. As a result,
+/// if either is violated, then the `Drop` impl on `DiagnosticGuard` will
+/// panic.
+pub(super) struct DiagnosticGuard<'sink> {
+    file: File,
+
+    sink: &'sink std::cell::RefCell<TypeCheckDiagnostics>,
```

Comment:
What goes wrong when you use `&mut TypeCheckDiagnostics` here?

It might be worth a brief comment saying why interior mutability is being used here.